r/R/table.R

```diff
@@ -254,6 +257,68 @@ names.Table <- function(x) x$ColumnNames()
 #' @export
 `[[.Table` <- `[[.RecordBatch`
 
+#' @export
+`[[<-.Table` <- function(x, i, value) {
+  if (!is.character(i) & !is.numeric(i)) {
+    stop("'i' must be character or numeric, not ", class(i), call. = FALSE)
+  } else if (is.na(i)) {
+    # Catch if a NA_character or NA_integer is passed. These are caught elsewhere
+    # in cpp (i.e. _arrow_RecordBatch__column_name)
+    # TODO: figure out if catching in cpp like ^^^ is preferred
```

Comment:
If we get a useful error message from the cpp layer, great, use that. Let the static(ish) typing do the work.

Reply:
It's a mixture, for `NA_integer_` and `NA_real_` the error is the previous "missing value where TRUE/FALSE needed" with `NA_character_` though it removes/creates/replaced a column with the name ``NA``. So I would vote to keep the special casing for now and try catching in cpp or with a helper function we can use across `[[` and `[` methods later.

Reply:
I suspect that error comes from https://github.com/apache/arrow/pull/8579/files#diff-242f7f20fb4999f3bbe719d6977d2580231b89334326b588d46e9236c2e28fcdR285, i.e. before we hit cpp.